lib/Onyx.js

```diff
@@ -712,12 +713,14 @@ function clear() {
     return getAllKeys()
         .then((keys) => {
             _.each(keys, (key) => {
-                keyChanged(key, null);
-                cache.set(key, null);
+                const resetValue = lodashGet(defaultKeyStates, key, null);
+                cache.set(key, resetValue);
+
+                // Optimistically inform subscribers on the next tick
```

Comment:
+1

Reply:
See https://github.com/Expensify/react-native-onyx/pull/129#issuecomment-1136467520

Reply:
> Not sure I get it. If we are calling a `merge()` then a `clear()` in the same tick then wouldn't the correct behavior be for the `clear()` to cancel the `merge()`... ?

Without this, `clear()` runs before `merge()` since `merge()` is delayed by `get(key)` and `set()`. What are you thoughts on that?
https://github.com/Expensify/react-native-onyx/blob/5478ec947dbbbfaaaa29923c90d47314117ff83d/lib/Onyx.js#L673-L688

Reply:
Sorry, not seeing the issue. `Onyx.clear()` runs before `Onyx.merge()` when we do this?

```js
Onyx.merge('something', {test: 'wat'});
Onyx.clear();
```

>clear() runs before merge() since merge() is delayed by get(key) and set()

Can we maybe break down this explanation a bit further ?

Reply:
Alright, ignore my last explanation. Here is what I have found from logging.
```javascript
Onyx.merge('something', {test: 'wat'});
Onyx.clear();
```
Without the change in question, `keyChanged()` from `Onyx.set()` called via `Onyx.merge()` is called **after** `keyChanged()` from `Onyx.clear()`. This happens because in `Onyx.merge()` we call `Onyx.get()` which finishes before `Onyx.getAllKeys()`. Back in `Onyx.merge()` once we get the key we call `Onyx.set()`. We set the cache and then we call `keyChanged()` on the next tick and we call `Storage.setItem()` for the merged key. Next we go back to `Onyx.clear()` where we loop through all the keys we got and we set the cache and call `keyChanged()` on each key. Immediately after that `keyChanged()` runs with the merged value. Then we go back to `Onyx.clear()` and we call `Storage.clear()`.

That is the saga of merge then clear, and is how the Onyx value is the merged value while the cache and storage are the default value. Calling `keyChanged()` on the next tick fixes this problem.

Reply:
>This happens because in Onyx.merge() we call Onyx.get() which finishes before Onyx.getAllKeys() called from Onyx.clear(). 

Ok so the answer is that we have to do this because we are doing this in `set()`:

https://github.com/Expensify/react-native-onyx/blob/b61c904a55d5410679e08a9fa6bd1bad10f1f894/lib/Onyx.js#L552-L553

We need to do it here as well otherwise a call to `Onyx.clear()` will call `keysChanged()` first despite running later.

That change is pretty old but I think we are probably scared to change it at this point :D 

I would suggest we change that usage to something like `notifySubscribersOnNextTick()` or something and refactor the few usages we have where we are doing `Promise.resolve().then(() => keyChanged())` - it might help contextualize better why we need to do this in some but not all places.

Reply:
I still think we should improve this later but the context is here so I'm good with this.

Reply:
> I would suggest we change that usage to something like `notifySubscribersOnNextTick()` or something and refactor

Do you want me to do that now, in a follow up, or let someone else do that?